index.js

```diff
@@ -1,301 +1,318 @@
+   * and return Stream instance ready to be writen on
+   * @param path
+   * @param options
+   */
+  constructor(path, options) {
+    options = Object.assign({}, DEFAULT_OPTIONS, options)
+    if (options.flags.indexOf('a') === -1) {
+      throw new Error('RotateStream must be opened in append mode')
     }
-
-    if ( options.flags.indexOf( 'a' ) == -1 ) {
-        throw new Error( 'RotateStream must be opened in append mode' );
-    }
-
-    fs.WriteStream.call( this, path, options );
-
-    this.size = isNaN( options.size )
-        ? bytes( options.size ) : options.size;
-
-    this.suffix = options.suffix || '';
-
-    var that = this;
-    this.on( 'open', function () {
-        fs.fstat( this.fd, function ( err, stats ) {
-            if ( err ) {
-                that.emit( 'error', err );
-            } else {
-                that.ino = stats.ino;
-            }
-        })
-    })
-
-    this._compressq = [];
-    if ( options.gzip ) {
-        this.on( 'rotated', function ( _, path ) {
-            process.nextTick( this._compress.bind( this, path ) );
-        });
+    super(path, options)
+
+    this.compress = options.gzip
+    this.size = isNaN(options.size) ? bytes(options.size) : options.size
+    this.suffix = options.suffix
+    if (options.upload) {
+      if (!aws) {
+        throw new Error('The aws-sdk module isn\'t installed')
+      }
+
+      this.toUpload = true
+      this._s3 = _getS3(options.upload)
     }
-
-    this._uploadq = [];
-    if ( options.upload ) {
-        if ( !aws ) {
-            throw new Error( 'The aws-sdk module isn\'t installed' )
-        }
-
-        var comps = options.upload.match( /s3\:\/\/((.*)\@)?([^\/]+)\/(.*)/ );
-
-        if ( !comps ) {
-            throw new Error( 'Malformed S3 upload path' );
-        }
-
-        var credentials = ( comps[ 2 ] || '' ).split( ':' );
-        var bucket = comps[ 3 ];
-        var prefix = comps[ 4 ] || '';
-
-        if ( !bucket ) {
-            throw new Error( 'S3 Upload bucket is not defined' );
-        }
-
-        var s3 = {
-            signatureVersion: 'v4'
-        }
-
-        if ( credentials[ 0 ] ) {
-            s3.accessKeyId = credentials[ 0 ];
-        }
-
-        if ( credentials[ 1 ] ) {
-            s3.secretAccessKey = credentials[ 1 ];
+  }
+
+  /**
+   * @private
+   */
+  _openAsync() {
+    let self = this
+    return fs.openAsync(this.path, this.flags, this.mode)
+      .then((fd) => {
+        self.fd = fd
+        return fs.fstatAsync(fd)
+      })
+      .then((stats) => {
+        self.ino = stats.ino
+      })
+      .catch((err) => {
+        if (self.autoClose) {
+          self.destroy()
         }
+        throw err
+      })
+  }
+
+  /**
+   * @private
+   */
+  _reopen() {
+    let self = this
+
+    this.bytesWritten = 0
+    return fs.closeAsync(self.fd)
+      .then(() => {
+        self.closed = self.destroyed = false
+        self.fd = null
+      })
+      .bind(self).then(self._openAsync)
```

Comment:
For future reference, since you're already binding, you don't need `let self = this`. You can just bind earlier. Not a requested change, just a comment.